# Lista Aberta do Módulo 3 - Linear na Prática

## Predizendo a glicose do sangue - Diabetes Dataset

Neste trabalho você deverá usa uma regressão linear para o problema de **predição da glicose do sangue**. Para isso, usaremos o dataset de diabetes de Stanford. Notem que as variáveis já estão normalizadas.

Você não precisa implementar os métodos, já que estão disponíveis na biblioteca [statsmodels](https://www.statsmodels.org/stable/index.html) da linguagem Python. Se necessário, pode fazer mais importações de bibliotecas.

## Importando módulos e baixando dataset

In [ ]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf

from sklearn.datasets import load_diabetes

In [ ]:
diabetes = load_diabetes()
df = pd.DataFrame(diabetes.data,columns =[diabetes.feature_names])

In [ ]:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

In [ ]:
df

age       sex       bmi        bp        s1        s2        s3  \
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
..        ...       ...       ...       ...       ...       ...       ...   
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

           s4        s5        s6  
0   -0.002592  0.019907 -0.017646  
1   -0.039493 -0.068332 -0.092204  
2   -0.002592  0.002861 -0.025930  
3    0.034309  0.022688 -0.009362  
4   -0.002592 -0.031988 -0.046641  
..        ...       ...       ...  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018114  0.044485  
439 -0.011080 -0.046883  0.015491  
440  0.026560  0.044529 -0.025930  
441 -0.039493 -0.004222  0.003064  

[442 rows x 10 columns]

In [ ]:
# Variavel que queremos predizer
TARGET = "s6"

Se tiver dúvidas sobre a validação cruzada, recomendo :https://machinelearningmastery.com/k-fold-cross-validation/ . Busque também outros materiais, o importante é você conhecer e saber utilizar a validação cruzada, que é bem mais robusta que a simples divisão em treino e teste.

# Exercício 01:

1. Execute a regressão em statsmodels para prever a variável 6. Para tal, use como exemplo o tutorial [daqui](https://www.statsmodels.org/stable/regression.html). Use apenas as variáveis age, sex, bmi e bp como preditoras.
1. Interprete os intervalos de confiança das variáveis. Existe alguma que poderia ser eliminada? Se sim, qual?

In [ ]:
variavel_prevista = df[TARGET]
varivaies_regressao = df[['age', 'sex', 'bmi', 'bp']]

modelo = sm.OLS(variavel_prevista, varivaies_regressao)
resultado = modelo.fit()

print(resultado.summary())

                                 OLS Regression Results                                
Dep. Variable:                     s6   R-squared (uncentered):                   0.256
Model:                            OLS   Adj. R-squared (uncentered):              0.249
Method:                 Least Squares   F-statistic:                              37.62
Date:                Sun, 19 Nov 2023   Prob (F-statistic):                    4.63e-27
Time:                        16:09:10   Log-Likelihood:                          784.28
No. Observations:                 442   AIC:                                     -1561.
Df Residuals:                     438   BIC:                                     -1544.
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

**Resposta 2:** A variável "sex" é aquela que tem menor intervalo de confiança, além de ter o menor valor "t". Assim, essa característica é a que menos contribui para a regressão.

Ao removê-la, obtemos:

In [ ]:
modelo = sm.OLS(df[TARGET], df[['age', 'bmi', 'bp']])
resultado = modelo.fit()

print(resultado.summary())

                                 OLS Regression Results                                
Dep. Variable:                     s6   R-squared (uncentered):                   0.245
Model:                            OLS   Adj. R-squared (uncentered):              0.240
Method:                 Least Squares   F-statistic:                              47.51
Date:                Sun, 19 Nov 2023   Prob (F-statistic):                    1.31e-26
Time:                        16:09:10   Log-Likelihood:                          781.15
No. Observations:                 442   AIC:                                     -1556.
Df Residuals:                     439   BIC:                                     -1544.
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

Como houve uma aproximação do "R-squared" e do "Adj. R-squared", é possível notar que houve uma melhora no modelo.

# Exercício 02:

1. Execute DUAS regressões para prever a quantidade de crimes violentos (violent) e quantidade de assassinatos (murder). Não use crimes violentos para prever assassinatos, nem assassinatos para precer crimes violentos. Use apenas as outras variáveis nos dois casos.
1. Interprete os intervalos de confiança das variáveis. Existe alguma que poderia ser eliminada? Se sim, qual?

In [ ]:
crime_data = sm.datasets.statecrime.load()
df = crime_data.data
df

violent  murder  hs_grad  poverty  single  white   urban
state                                                                         
Alabama                 459.9     7.1     82.1     17.5    29.0   70.0   48.65
Alaska                  632.6     3.2     91.4      9.0    25.5   68.3   44.46
Arizona                 423.2     5.5     84.2     16.5    25.7   80.0   80.07
Arkansas                530.3     6.3     82.4     18.8    26.3   78.4   39.54
California              473.4     5.4     80.6     14.2    27.8   62.7   89.73
Colorado                340.9     3.2     89.3     12.9    21.4   84.6   76.86
Connecticut             300.5     3.0     88.6      9.4    25.0   79.1   84.83
Delaware                645.1     4.6     87.4     10.8    27.6   71.9   68.71
District of Columbia   1348.9    24.2     87.1     18.4    48.0   38.7  100.00
Florida                 612.6     5.5     85.3     14.9    26.6   76.9   87.44
Georgia                 432.6     6.0     83.9     16.5    29.3   61.9   65.38
Hawaii                  274.1     1.8     90.4     10.4    26.3   26.9   71.46
Idaho                   238.5     1.5     88.4     14.3    19.0   92.3   50.51
Illinois                618.2     8.4     86.4     13.3    26.0   72.5   79.97
Indiana                 366.4     5.3     86.6     14.4    24.5   85.7   59.17
Iowa                    294.5     1.3     90.5     11.8    20.3   92.3   41.66
Kansas                  412.0     4.7     89.7     13.4    22.8   86.3   50.17
Kentucky                265.5     4.3     81.7     18.6    25.4   88.8   40.99
Louisiana               628.4    12.3     82.2     17.3    31.4   63.7   61.33
Maine                   119.9     2.0     90.2     12.3    22.0   94.9   26.21
Maryland                590.0     7.7     88.2      9.1    27.3   60.2   83.53
Massachusetts           465.6     2.7     89.0     10.3    25.0   82.4   90.30
Michigan                504.4     6.3     87.9     16.2    25.6   79.9   66.37
Minnesota               214.2     1.5     91.5     11.0    20.2   87.4   58.00
Mississippi             306.7     6.9     80.4     21.9    32.8   59.6   27.62
Missouri                500.3     6.6     86.8     14.6    25.3   83.9   56.61
Montana                 283.9     3.2     90.8     15.1    20.3   89.4   26.49
Nebraska                305.5     2.5     89.8     12.3    20.9   88.1   53.78
Nevada                  704.6     5.9     83.9     12.4    28.5   76.2   86.51
New Hampshire           169.5     0.9     91.3      8.5    19.5   94.5   47.34
New Jersey              311.3     3.7     87.4      9.4    25.8   70.7   92.24
New Mexico              652.8    10.0     82.8     18.0    29.1   72.5   53.75
New York                385.5     4.0     84.7     14.2    30.2   67.4   82.66
North Carolina          414.0     5.4     84.3     16.3    26.3   70.5   54.88
North Dakota            223.6     2.0     90.1     11.7    18.2   90.2   40.00
Ohio                    358.1     5.0     87.6     15.2    26.3   84.0   65.31
Oklahoma                510.4     6.5     85.6     16.2    25.9   75.4   45.79
Oregon                  261.2     2.3     89.1     14.3    22.7   85.6   62.47
Pennsylvania            388.9     5.4     87.9     12.5    24.5   83.5   70.68
Rhode Island            254.3     3.0     84.7     11.5    27.3   82.6   90.46
South Carolina          675.1     6.7     83.6     17.1    28.4   67.6   55.78
South Dakota            201.0     3.6     89.9     14.2    20.8   86.3   29.92
Tennessee               666.0     7.4     83.1     17.1    26.3   79.1   54.38
Texas                   491.4     5.4     79.9     17.2    27.6   73.8   75.35
Utah                    216.2     1.4     90.4     11.5    17.9   89.3   81.17
Vermont                 135.1     1.3     91.0     11.4    21.3   95.8   17.38
Virginia                230.0     4.7     86.6     10.5    24.0   70.4   69.79
Washington              338.3     2.8     89.7     12.3    22.2   80.2   74.97
West Virginia           331.2     4.9     82.8     17.7    23.3   94.3  

## Regressão para Crimes Violentos

Partindo da suposição que todas as variáveis podem influênciar de forma significitiva a quantidade de crimes violentos em um determinado estado, temos a seguinte regressão:

In [ ]:
variavel_prevista = df['violent']
varivaies_regressao = df[['hs_grad', 'poverty', 'white', 'urban', 'single']]

modelo = sm.OLS(variavel_prevista, varivaies_regressao)
resultado = modelo.fit()

print(resultado.summary())

                                 OLS Regression Results                                
Dep. Variable:                violent   R-squared (uncentered):                   0.930
Model:                            OLS   Adj. R-squared (uncentered):              0.922
Method:                 Least Squares   F-statistic:                              122.4
Date:                Sun, 19 Nov 2023   Prob (F-statistic):                    2.15e-25
Time:                        16:09:10   Log-Likelihood:                         -317.23
No. Observations:                  51   AIC:                                      644.5
Df Residuals:                      46   BIC:                                      654.1
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

É possível notar que o intervalo de confiânça certas variáveis está alternando entre valores positivos e negativos e próximos de zero, como das variáveis "white" e "urban".

Além dos seus valores "t" serem muito proximos de 0, significando que não há uma influência delas sobre a variável dependente que não possa ser explicada pelo acaso.

A variável "poverty" também levanta uma certa confusão, pois ela indicaria que quanto maior a taxa de pobreza menor seria a quantidade de crimes violentos. Além disso, o seu valor "t" também é baixo. Então, também podemos descartá-la.

Dessa forma, podemos desconsiderar essas variáveis e obter a nova regressão:

In [ ]:
variavel_prevista = df['violent']
varivaies_regressao = df[['hs_grad', 'single']]

modelo = sm.OLS(variavel_prevista, varivaies_regressao)
resultado = modelo.fit()

print(resultado.summary())

                                 OLS Regression Results                                
Dep. Variable:                violent   R-squared (uncentered):                   0.928
Model:                            OLS   Adj. R-squared (uncentered):              0.925
Method:                 Least Squares   F-statistic:                              313.9
Date:                Sun, 19 Nov 2023   Prob (F-statistic):                    1.16e-28
Time:                        16:09:10   Log-Likelihood:                         -318.13
No. Observations:                  51   AIC:                                      640.3
Df Residuals:                      49   BIC:                                      644.1
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

Removendo essas variáveis não significativas, obtemos uma nova regressão com um "R-Squared" 0.928, bem próximo do anterior. Isso comprova ainda mais que elas não eram relevantes para a regressão.

Além disso, a diferênça entre o "Adj. R-squared" e "R-squared" reduziu. Isso é outro indício que aquelas variáveis poderiam ser removidas.

Interpretando a regressão, vemos que quanto maior a taxa de escolaridade (quantidade de pessoas com graduação), menor é a chance de ocorrer crimes violentos. Por outro lado, quanto maior a taxa de pessoas solteiras, maior será essa quantidade.

## Regressão para Assassinatos


Utilizando a mesma lógica para a quantidade de Assassinatos, temos a regressão abaixo considerando que todas as variáveis são relevantes para a variável dependente.

In [ ]:
variavel_prevista = df['murder']
varivaies_regressao = df[['hs_grad', 'poverty', 'white', 'urban', 'single']]

modelo = sm.OLS(variavel_prevista, varivaies_regressao)
resultado = modelo.fit()

print(resultado.summary())

                                 OLS Regression Results                                
Dep. Variable:                 murder   R-squared (uncentered):                   0.916
Model:                            OLS   Adj. R-squared (uncentered):              0.906
Method:                 Least Squares   F-statistic:                              99.75
Date:                Sun, 19 Nov 2023   Prob (F-statistic):                    1.62e-23
Time:                        16:09:10   Log-Likelihood:                         -101.45
No. Observations:                  51   AIC:                                      212.9
Df Residuals:                      46   BIC:                                      222.5
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

Novamente, é possível notar intervalos de confiânça pequenos e próximos de zero, além de um valor "t" baixo para as variáveis "poverty", "urban" e "white". Além disso, os seus intervalos de confiânça estão variando entre os valores positivos e negativos e estão próximos de 0.

Assim, podemos removê-las, gerando uma nova regressão:

In [ ]:
variavel_prevista = df['murder']
varivaies_regressao = df[['hs_grad', 'single']]

modelo = sm.OLS(variavel_prevista, varivaies_regressao)
resultado = modelo.fit()

print(resultado.summary())

                                 OLS Regression Results                                
Dep. Variable:                 murder   R-squared (uncentered):                   0.913
Model:                            OLS   Adj. R-squared (uncentered):              0.910
Method:                 Least Squares   F-statistic:                              257.6
Date:                Sun, 19 Nov 2023   Prob (F-statistic):                    9.97e-27
Time:                        16:09:10   Log-Likelihood:                         -102.16
No. Observations:                  51   AIC:                                      208.3
Df Residuals:                      49   BIC:                                      212.2
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

Assim como antes, obtemos uma regressão com um "R-squared" de 0.913, que é bem próximo da regressão inicial.

A interpretação da regressão é a mesma da anterior. Quanto maior o nível de escolaridade, menos casos de assassinatos. Além disso, quanto mais pessoas solteiras, maior a chance de assassinatos.